synapse/handlers/register.py

```diff
@@ -222,14 +223,19 @@ def register(
         fake_requester = create_requester(user_id)
 
         # try to create the room if we're the first user on the server
+        should_auto_create_rooms = False
         if self.hs.config.autocreate_auto_join_rooms:
             count = yield self.store.count_all_users()
-            auto_create_rooms = count == 1
+            should_auto_create_rooms = count == 1
 
         for r in self.hs.config.auto_join_rooms:
             try:
-                if auto_create_rooms and RoomAlias.is_valid(r):
+                if should_auto_create_rooms:
                     room_creation_handler = self.hs.get_room_creation_handler()
+                    if self.hs.hostname != RoomAlias.from_string(r).domain:
+                        raise ConfigError(
```

Comment:
... and if we can't check it during startup, it's probably more appropriate to warn and move on than to 500 reject the whole registration.